### DISCUSSION TITLE
Httpd docker image in docker hub

### DISCUSSION THREAD

**Discussion by fong1999** [OpenIDC/mod_auth_openidc#1131] [Q&A]
Hi! I have some general questions regarding docker images.

Which team is responsible for building the docker image? How often will that being pushed to Docker hub?

https://hub.docker.com/_/httpd

Thank you!

**Comment by zandbelt** [comment]
this is a separate project, not related to mod_auth_openidc at all so I wouldn't know

**Reply by fong1999** [reply]
@zandbelt Hi! Do they get the latest release from https://github.com/OpenIDC/mod_auth_openidc/releases and build the docker image? Do you know which team is that? I want to reach out to them. Thanks!

**Reply by hmoffatt** [reply]
That's the official Apache httpd image. It looks like you are adding mod_auth_openidc, based on Dockerfile snippets you posted previously?

**Reply by fong1999** [reply]
@hmoffatt Hi! Thanks for your reply!  Correct. We are using So using 2.4.58-debian or 2.4.57-debian will not make much difference, it is the C library that we are using along with the httpd server, right? 

**We are still seeing below inside of the container when I am testing to use 2.4.58-debian-ff. Not sure why it shows 2.4.57-debian. Any ideas? Thanks!**

```
>kubectl exec -it adw-aws-dev-poc-deployment-758569658f-t6gbf sh -n inside-dev
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.

# apache2 -v
Server version: Apache/2.4.57 (Debian)
Server built:   2023-04-13T03:26:51
#  apt show libcjose0 libapache2-mod-auth-openidc libhiredis0.14 | grep Version

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Version: 0.6.2.1-1+deb12u1
Version: 2.4.12.3-2
Version: 0.14.1-3
# exit
```

The issue is still the same as using 2.4.57-debian. which gives us "500 error" (I described in my other discussion)

```
 oidc_authenticate_user: the URL scheme (https) of the configured OIDCRedirectURI does not match the URL scheme of the URL being accessed (http): the "state" and "session" cookies will not be shared between the two!
```

We reverted it back to use 2.4.56-debian-ff again, which does not show the error above, and our app is running fine.

```
# apache2 -v
Server version: Apache/2.4.56 (Debian)
Server built:   2023-04-02T03:06:01
#  apt show libcjose0 libapache2-mod-auth-openidc libhiredis0.14 | grep Version

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Version: 0.6.1+dfsg1-1+deb11u1
Version: 2.4.9.4-0+deb11u3
Version: 0.14.1-1
```
If this is the cause, 

2.4.57-debian (does not work)
```
Version: 2.4.12.3-2
```
2.4.56-debian (does work)
```
Version: 2.4.9.4-0+deb11u3
```

Is it possible or Is there a way to customize by pulling Version: 2.4.9.4-0+deb11u3 with combination using 2.4.57 or later debian till a new version of httpd works with newer version of mod_aut_openidc library? Any ideas? Thanks much!

**Reply by hmoffatt** [reply]
Sure, there is httpd:2.4.57-bullseye on dockerhub and you can just install libapache2-mod-auth-openidc on that, which will be `2.4.9.4-0+deb11u3` as that's what is in Debian bullseye.

Apache 2.4.58 just came out though and there are no bullseye images for that, so you will have to figure out the problem eventually.

**Reply by fong1999** [reply]
@hmoffatt Hi! Thanks for the info! I am going try that - **2.4.57-bullseye** next. I will keep you posted.  Thanks!

I was testing the new image: 

**Dockerfile**

```
FROM xxxxx/library/xxxx:2.4.58-debian
```

I noticed that when I verify inside of the container, I see below:

```
>kubectl exec -it adw-aws-dev-deployment-75768c89c9-7f76t sh -n inside-dev
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
Defaulted container "adw-aws-dev" out of: adw-aws-dev, stunnel
# httpd -v
Server version: Apache/2.4.58 (Unix)
Server built:   Oct 19 2023 19:20:54
# apache2 -v
Server version: Apache/2.4.57 (Debian)
Server built:   2023-04-13T03:26:51
```

Any ideas why it shows different versions? Thanks!

**Comment by fong1999** [comment]
@hmoffatt Hi! Unfortunatlly, `httpd:2.4.57-bullseye` hae CVEs, we cannot use it. Do you recall which version that this issue was solved? We were having this issue in the past, but it didn't stop the app from running, which means that we do not get 500 error. After using 2.4.56 (maybe the other versions...), this issue went away for a while. Now it came back with giving us 500 error. If you remember which version that might fix it, I might try that version of mod_auth_oidc and see if it helps. All we are trying is do find the version that works with our app, and does not have any CVEs. : ) Thanks much!

```
 oidc_authenticate_user: the URL scheme (https) of the configured OIDCRedirectURI does not match the URL scheme of the URL being accessed (http): the "state" and "session" cookies will not be shared between the two!
```

**Reply by hmoffatt** [reply]
You can probably take the httpd:2.4.58 image and compile the old version of mod_auth_openidc from source in your Dockerfile.

Do you have some other HTTP server in front of Apache, like a load balancer?

**Reply by fong1999** [reply]
@hmoffatt Could you please be a little more specific? When I use 2.4.58-debian in my Dockerfile, then apt install those three c libraries including mod_auth_openidc, it builds the docker image but with those specified versions as shown in my previous message. Do you mean build an image that way or? 

We use Apache server for the web app. In AWS, we have  ingress and ALB….

To remediate the CVE issue, I am thinking instead of waiting for docker hub to have a new httpd image that works with your mod_auth_openidc latest release 2.14.4. I can try to build it to test it out. Is it easy to do? How do I do that? 

Thanks!

**Reply by hmoffatt** [reply]
Is your ALB connecting to your Apache with HTTP or HTTPS?

I don't know what `2.4.58-debian` is, there is no such tag for httpd on hub.docker.com.

**Reply by fong1999** [reply]
@hmoffatt Hi! Sorry for the confusion. I think `-debain` might be what our company middleware team added to it. Do you mean using 'Dockerfile` (debian:bookworm) that is in Docker hub httpd site to build one and test out? Does mod_auth_openidc **2.14.4** work with any version of httpd docker image that build in docker hub as long as it is debian:bookworm if we use debian distribution?

https://github.com/docker-library/httpd/blob/338f9a4b1e36e09ad2733f36f19e31157f5c959c/2.4/Dockerfile

I guess my question is, How do I build a httpd image that works with mod_auth_openidc latest release? e.g. 2.14.4? As current 2.4.58 requires `2.4.12.3-2` of mod_auth_openidc... I want to test higher version of it. Or like you said, downgrade it to the version ` 2.4.9.4-0+deb11u3` that works with our app to remediate the CVE as short term solution? 

Below command is what we use in our Dockerfile to build our app docker image. I could specify the version to use `=2.4.9.4-0+deb11u3` and see. 

```
# Testing again
FROM ntr.nwie.net/library/httpd:2.4.58-debian

# Install mod_auth_openidc module for authentication
RUN apt-get update -y && \
    apt install -y libcjose0 libapache2-mod-auth-openidc=2.4.9.4-0+deb11u3 libhiredis0.14 && \
	apt clean
```

Thanks!

**Reply by fong1999** [reply]
@hmoffatt However, I am getting error message during docker image build in the pipleline:

```
E: Version '2.4.9.4-0+deb11u3' for 'libapache2-mod-auth-openidc' was not found
The command '/bin/sh -c apt-get update -y &&     apt install -y libcjose0 libapache2-mod-auth-openidc=2.4.9.4-0+deb11u3 libhiredis0.14 && 	apt clean' returned a non-zero code: 100
```

I think it looks for the compatible version `2.4.12.3-2`:

```
# httpd -v
Server version: Apache/2.4.58 (Unix)
Server built:   Oct 19 2023 19:20:54
# apt show libcjose0 libapache2-mod-auth-openidc libhiredis0.14 | grep Version

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Version: 0.6.2.1-1+deb12u1
Version: 2.4.12.3-2
Version: 0.14.1-3
```

**### Do you know how httpd builds their image knowing mod_auth_openidc has a newer release? Thanks!**

At meanwhile, let me try this:

```
ntr.nwie.net/docker.io/httpd:2.4.57-bullseye

# Install mod_auth_openidc module for authentication
RUN apt-get update -y && \
    apt install -y libcjose0 libapache2-mod-auth-openidc=2.4.9.4-0+deb11u3 libhiredis0.14 && \
	apt clean
```

Hope this will work. brb. Thanks!

**Reply by fong1999** [reply]
@hmoffatt  Update: Using bullseye seems worked. However, 2.4.57-bulleye has CVEs, we cannot use it. 

`
Get:12 http://deb.debian.org/debian bullseye/main amd64 libapache2-mod-auth-openidc amd64 2.4.9.4-0+deb11u3 [181 kB]
`

**Do you know how httpd builds their image when mod_auth_openidc has a newer release?** I was checking their github repo, but didn't find a place to submit a question... 

https://github.com/docker-library/httpd

Thanks!

**Reply by fong1999** [reply]
@hmoffatt Hi! 

More updates: 

I am working with one of our tech consultant resources. We are building our docker image using the latest mod_auth_openidc release 2.14.4 with 2.4.x (should be pulling 2.4.58). I will keep you posted. Thanks!

**Reply by hmoffatt** [reply]
httpd has nothing to do with this module, it is just Debian + Apache, from the Apache developers. As it's Debian underneath you can install whatever libapache2-mod-auth-openidc package is in Debian. For bookworm, which the 2.4.58 package is using, that is 2.4.12.3.

**Reply by fong1999** [reply]
@hmoffatt Hi! With my colleague's help, we were able to build our docker image with 2.4.58 (bookwarm)  with mod_auth_openidc 2.4.14.4. But we are still getting 500 error.  We are going to do more research on what could be the change in httpd that is causing this or we need to adjust out httpd.conf setting. I will keep you posted. Thanks!

**Reply by fong1999** [reply]
@hmoffatt Hi! With the 2.4.56 that works with our app, but I noticed we are getting **a lot of errors** in our log. Even though our app is working, but these logs make me concern. Please advise. Thanks!

For example,

```
2023-07-28 05:06:27.814	[Fri Jul 28 05:06:27.814125 2023] [auth_openidc:error] [pid 1631:tid 139962884273920] [client 10.0.12.37:41425] oidc_restore_proto_state: calculated state from cookie does not match state parameter passed back in URL: "ko-ntGyQoYK9521aJBSLj8lSg" != "n9xzmM2K3cEY1_UFiwdxIIgi8", referer: https://login.microsoftonline.com/
2023-07-28 05:06:27.814	[Fri Jul 28 05:06:27.814144 2023] [auth_openidc:error] [pid 1631:tid 139962884273920] [client 10.0.12.37:41425] oidc_authorization_response_match_state: unable to restore state, referer: https://login.microsoftonline.com/
2023-07-28 05:06:27.814	[Fri Jul 28 05:06:27.814148 2023] [auth_openidc:error] [pid 1631:tid 139962884273920] [client 10.0.12.37:41425] oidc_handle_authorization_response: invalid authorization response state and no default SSO URL is set, sending an error..., referer: https://login.microsoftonline.com/
2023-07-28 07:34:48.577	[Fri Jul 28 07:34:48.577695 2023] [auth_openidc:error] [pid 220:tid 139668091819776] [client 10.0.10.165:9909] oidc_restore_proto_state: calculated state from cookie does not match state parameter passed back in URL: "ZCXGPF6yVTMzHLrt3RpJhsoUk" != "AqcUChCkX6qgn0bZUAvO2hxeY", referer: https://login.microsoftonline.com/
2023-07-28 07:34:48.577	[Fri Jul 28 07:34:48.577713 2023] [auth_openidc:error] [pid 220:tid 139668091819776] [client 10.0.10.165:9909] oidc_authorization_response_match_state: unable to restore state, referer: https://login.microsoftonline.com/
2023-07-28 07:34:48.577	[Fri Jul 28 07:34:48.577717 2023] [auth_openidc:error] [pid 220:tid 139668091819776] [client 10.0.10.165:9909] oidc_handle_authorization_response: invalid authorization response state and no default SSO URL is set, sending an error..., referer: https://login.microsoftonline.com/
2023-07-28 12:38:56.290	[Fri Jul 28 12:38:56.290354 2023] [auth_openidc:error] [pid 1631:tid 139963362428672] [client 10.0.10.124:22004] oidc_restore_proto_state: calculated state from cookie does not match state parameter passed back in URL: "-3UmvuM_7WlZJGcZ76BnwuY1E" != "y5R7dlTLWn45tI5AbxILQHMY", referer: https://login.microsoftonline.com/
2023-07-28 12:38:56.290	[Fri Jul 28 12:38:56.290373 2023] [auth_openidc:error] [pid 1631:tid 139963362428672] [client 10.0.10.124:22004] oidc_authorization_response_match_state: unable to restore state, referer: https://login.microsoftonline.com/
2023-07-28 12:38:56.290	[Fri Jul 28 12:38:56.290377 2023] [auth_openidc:error] [pid 1631:tid 139963362428672] [client 10.0.10.124:22004] oidc_handle_authorization_response: invalid authorization response state and no default SSO URL is set, sending an error..., referer: https://login.microsoftonline.com/
2023-07-28 16:51:49.706	[Fri Jul 28 16:51:49.706312 2023] [auth_openidc:error] [pid 220:tid 139668318467840] [client 10.0.0.53:15488] oidc_restore_proto_state: no "mod_auth_openidc_state_VhYn7FU9TIqstPftNH_B2io_c" state cookie found: check domain and samesite cookie settings
2023-07-28 16:51:49.706	[Fri Jul 28 16:51:49.706330 2023] [auth_openidc:error] [pid 220:tid 139668318467840] [client 10.0.0.53:15488] oidc_authorization_response_match_state: unable to restore state
```

**Reply by hmoffatt** [reply]
The cookies are missing, which is exactly what it is warning you about.

**Reply by fong1999** [reply]
@hmoffatt Thanks for the info! We are going to try some alternative today. I will keep you posted. Thanks!

**Reply by fong1999** [reply]
@hmoffatt Hi! What we built is to use 2.4.58 httpd image and the latest mod_auth_openidc 2.4.14.4 as shown below. 

```
FROM ntr.xxx/library/httpd:2.4

ARG MOD_AUTH_OPENIDC_VERSION=2.4.14.4

RUN \
    BUILD_DEPS=" \
        curl" \
    && RUNTIME_DEPS="\
        apache2-bin \
        libcjose0 \
        libhiredis0.14" \
    && MOD_AUTH_OPENIDC_URL=https://github.com/OpenIDC/mod_auth_openidc/releases/download/v"$MOD_AUTH_OPENIDC_VERSION"/libapache2-mod-auth-openidc_"$MOD_AUTH_OPENIDC_VERSION"-1.bookworm_amd64.deb \
    && apt update \
    && apt install -y --no-install-recommends \
        $BUILD_DEPS $RUNTIME_DEPS \
    && rm -r /var/lib/apt/lists/* \
    && curl -k --location "$MOD_AUTH_OPENIDC_URL" -o /tmp/mod_auth_openidc.deb \
    && dpkg -i /tmp/mod_auth_openidc.deb \
    && ln -s /usr/lib/apache2/modules/mod_auth_openidc.so /usr/local/apache2/modules/mod_auth_openidc.so \
    && apt purge -y --auto-remove \
        $BUILD_DEPS \
    && apt clean \
    && rm -rf \
        /tmp/* \
        /var/log/dpkg.log \
        /var/log/alternatives.log \
        /var/log/apt \
        /var/lib/apt/lists/*
```
```
# httpd -v
Server version: Apache/2.4.58 (Unix)
Server built:   Nov  1 2023 02:32:26
# apt show libcjose0 libapache2-mod-auth-openidc libhiredis0.14 | grep Version

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Version: 0.6.2.1-1+deb12u1
Version: 2.4.14.4-1.bookworm
Version: 0.14.1-3
```

However, when deploy to our environment, I see the same error - 500, and log message below.

```
[Thu Nov 02 00:39:01.757305 2023] [auth_openidc:error] [pid 11:tid 139729009882816] [client 10.0.3.117:53504] oidc_session_load_cache: cache backend failure for key 78591924-0a78-45ad-b383-776c03ea85db, referer: https://adw-aws-dev.apps.nwie.net/SearchResults
[Thu Nov 02 00:39:01.757314 2023] [auth_openidc:warn] [pid 11:tid 139729009882816] [client 10.0.3.117:53504] oidc_check_x_forwarded_hdr: header X-Forwarded-Host received but OIDCXForwardedHeaders not configured for it, referer: https://adw-aws-dev.xxx/SearchResults
[Thu Nov 02 00:39:01.757321 2023] [auth_openidc:warn] [pid 11:tid 139729009882816] [client 10.0.3.117:53504] oidc_check_x_forwarded_hdr: header X-Forwarded-Port received but OIDCXForwardedHeaders not configured for it, referer: https://adw-aws-dev.xxx/SearchResults
[Thu Nov 02 00:39:01.757324 2023] [auth_openidc:warn] [pid 11:tid 139729009882816] [client 10.0.3.117:53504] oidc_check_x_forwarded_hdr: header X-Forwarded-Proto received but OIDCXForwardedHeaders not configured for it, referer: https://adw-aws-dev.xxx/SearchResults
[Thu Nov 02 00:39:01.759238 2023] [auth_openidc:warn] [pid 93:tid 139729035060928] [client 10.0.3.117:53496] oidc_delete_oldest_state_cookies: deleting oldest state cookie: mod_auth_openidc_state_ZUXOZ6Ld8wbWRJ4L28f77A (time until expiry -60 seconds), referer: https://adw-aws-dev.xxx/SearchResults
[Thu Nov 02 00:39:01.759258 2023] [auth_openidc:warn] [pid 93:tid 139729035060928] [client 10.0.3.117:53496] oidc_delete_oldest_state_cookies: deleting oldest state cookie: mod_auth_openidc_state_vhkLpjDzyuxXVog16pPdzvL4 (time until expiry -60 seconds), referer: https://adw-aws-dev.xxx/SearchResults
[Thu Nov 02 00:39:01.759263 2023] [auth_openidc:warn] [pid 93:tid 139729035060928] [client 10.0.3.117:53496] oidc_delete_oldest_state_cookies: deleting oldest state cookie: mod_auth_openidc_state_-GG2c79xDysAw9zj7BlOFLSc (time until expiry -60 seconds), referer: https://adw-aws-dev.apps.nwie.net/SearchResults
[Thu Nov 02 00:39:01.759272 2023] [auth_openidc:error] [pid 93:tid 139729035060928] [client 10.0.3.117:53496] oidc_authenticate_user: the URL scheme (https) of the configured OIDCRedirectURI does not match the URL scheme of the URL being accessed (http): the "state" and "session" cookies will not be shared between the two!, referer: https://adw-aws-dev.xxx/SearchResults
```

It looks like 2.4.14.4 with httpd 2.4.58 do not work. I am going to try `2.4.9.4-0+deb11u3`, which I know it works with 2.4.56.  Is this the version? I don't see it in - https://github.com/OpenIDC/mod_auth_openidc/tags

Does this `2.4.9.4-0+deb11u3` work with **bookworm** or **bulleye**? I think you mentioned that it was with **bookworm**, right? Thanks!

**Reply by hmoffatt** [reply]
Your Dockerfile looks good. The old deb is built for bullseye (deb11u3), it may or may not work with bookworm, and I don't know what your httpd:2.4 is based on because the repository you refer to is private (ntr.nwie.net/library/httpd:2.4).

All of the warnings look like configuration errors though to me. Maybe @zandbelt can comment on the cache backend failure.

**Reply by zandbelt** [reply]
I suggest reading the error messages and act upon those by fixing the config first and then take it from there.
